File: xna/BlackjackXNA/BlackjackXNA/dealer.cs
/*
  Blackjack
  Copyright 2015 Sam Saint-Pettersen
  Released under the MIT/X11 License.

  C#/XNA implementation
*/
using System;
using System.Collections.Generic;
using Helper;

namespace BlackjackXNA
{
  /// <summary>
  /// Dealer implements the dealer for Blackjack.
  /// </summary>
  class Dealer 
  {
    private bool debug;
    private int index;
    private int pos;
    private List<string> cards;
    private List<int> values;
    private Cards gameCards;
  
    /// <summary>
    /// Constructor for dealer.
    /// </summary>
    /// <param name="debug">Enable debug messages?</param>
    /// <param name="gameCards">Game cards.</param>
    public Dealer(bool debug, Cards gameCards) 
    {
      this.debug = debug;
      this.index = 0;
      this.pos = 225;
      this.cards = new List<string>();
      this.values = new List<int>();
      this.gameCards = gameCards;
    }
  
    /// <summary>
    /// Calculate the total value of dealer's held cards.
    /// </summary>
    /// <returns>Total value of dealer's cards.</returns>
    public int CalcTotal() 
    {
      this.values.Sort(); //.OrderByDescending().toList();
      int total = 0;
      for(int i = 0; i < this.values.Count; i++) 
      {
        int v = this.values[i];
        if(v == 1) 
        {
          if((total + 11) <= 21) v = 11;
          else if((total + 11) > 21) v = 1;
        }
        total += v;
      }
      return total;
    }
  
    /// <summary>
    /// Dealer hits.
    /// </summary>
    /// <param name="cards">Game cards.</param>
    /// <returns>Dealer's drawn card.</returns>
    private Card Hit(Cards cards) 
    {
      this.index++;
      this.pos = 90;
      string card = cards.Draw();
      this.cards.Add(card);
      this.values.Add(cards.GetValue());
      Debugger.Emit(this.debug, "Dealer hits.");
      Debugger.Emit(this.debug, "Dealer gets " + card);
      return new Card(cards.GetImage(card), this.pos, 10);
    }
  
    /// <summary>
    /// Dealer stands.
    /// </summary>
    private void Stand() 
    {
      Debugger.Emit(this.debug, "Dealer stands.");
    }
  
    /// <summary>
    /// Dealer shuffles.
    /// </summary>
    public void Shuffle()
    {
        Debugger.Emit(this.debug, "----------------------------------------------------");
        Debugger.Emit(this.debug, "Dealer is shuffling cards....");
        Debugger.Emit(this.debug, "----------------------------------------------------");
        this.gameCards.Shuffle();
    }
  
    /// <summary>
    /// Dealer deals.
    /// </summary>
    /// <param name="cards">Game cards.</param>
    /// <returns>Player's cards.</returns>
    public string[] Deal(Cards cards) 
    {
      List<string> dealt = new List<string>();
      int i = 1;
      Debugger.Emit(this.debug, "----------------------------------------------------");
      Debugger.Emit(this.debug, "Dealer is dealing cards for a new game...");
      Debugger.Emit(this.debug, "----------------------------------------------------");
      while(i <= (2 * 2)) 
      {
        dealt.Add(cards.Draw() + " " + cards.GetValue());
        i++;
      }
      i = 0;
      while(i < 2) 
      {
        string[] cv = dealt[i].Split();
        this.cards.Add(cv[0]);
        this.values.Add(int.Parse(cv[1]));
        i++;
      }
      Debugger.Emit(this.debug, "\nDealer has:");
      Debugger.Emit(this.debug, String.Format("[**][{0}]", this.cards[1]));
      return new string[] { dealt[2], dealt[3] };
    }
  
    /// <summary>
    /// Determine if dealer has Blackjack.
    /// </summary>
    /// <returns>Does dealer have Blackjack?</returns>
    public bool HasBlackjack() 
    {
      bool blackjack = false;
      if(this.CalcTotal() == 21) 
      {
        Debugger.Emit(this.debug, "Dealer has Blackjack!");
        blackjack = true;
      }
      return blackjack;
    }
  
    /// <summary>
    /// Determine if dealer is bust.
    /// </summary>
    /// <returns>Is dealer bust?</returns>
    public bool IsBust() 
    {
      bool bust = false;
      if(this.CalcTotal() > 21)
      {
        Debugger.Emit(this.debug, "Dealer is bust!");
        bust = true;
      }
      return bust;
    }
  
    /// <summary>
    /// Dealer responds to player action (e.g. a hit or stand).
    /// </summary>
    /// <param name="cards">Game cards.</param>
    /// <returns>Cards returned.</returns>
    public List<Card> Respond(Cards cards) 
    {
      this.ShowCards();
      bool responding = true;
      List<Card> response_cards = new List<Card>();
      while(responding)
      {
        int total = 0;
        while(total <= 18) 
        {
          total = this.CalcTotal();
          if(total == 16)
          {
            Random rand = new Random();
            if(Math.Floor((double)rand.Next() * 6) >= 3) 
            {
              response_cards.Add(this.Hit(cards)); // Take risk.
            }
            else 
            {
              this.Stand(); // Play it safe.
            }
          }
          else if(total >= 17)
          {
            this.Stand();
            responding = false;
            break;
          }
          else
          {
            response_cards.Add(this.Hit(cards));
          }
        }
      }
      return response_cards;
    }
  
    /// <summary>
    /// Show dealer's cards.
    /// </summary>
    /// <returns>Total value of dealer's cards.</returns>
    public int ShowCards() 
    {
      this.index = 0;
      this.pos = 225;
      string cards = "";
      for(int i = 0; i < this.cards.Count; i++) 
      {
        cards += String.Format("[{0}]", this.cards[i]);
      }
      Debugger.Emit(this.debug, "\nDealer has:");
      Debugger.Emit(this.debug, String.Format("{0} --> {1}", cards, CalcTotal()));
      return CalcTotal();
    }
  
    /// <summary>
    /// Dealer receives cards.
    /// </summary>
    /// <param name="cards">Game cards.</param>
    /// <returns>Dealer's received cards.</returns>
    public List<Card> ReceiveCards(Cards cards) 
    {
      Card cardA = new Card(cards.GetImage("c"), this.pos, 10);
      this.pos += 90;
      Card cardB = new Card(cards.GetImage(this.cards[1]), this.pos, 10);
      this.index += 2;
      return new List<Card> { cardA, cardB };
    }
  
    /// <summary>
    /// Dealer reveals first card.
    /// </summary>
    /// <param name="cards">Game cards.</param>
    /// <returns>Revealed first card.</returns>
    public Card RevealFirstCard(Cards cards) 
    {
        return new Card(cards.GetImage(this.cards[0]), 225, 10);
    }
  }
}

File: html5/enchant.js_ts/dealer.ts
/*
	Blackjack
	Copyright 2015 Sam Saint-Pettersen
	Released under the MIT/X11 License.
	
	HTML5/enchant.js TypeScript implementation
*/

/// <reference path="enchant.d.ts" />
/// <reference path="sounds.ts" />
/// <reference path="card.ts" />
/// <reference path="cards.ts" />
/// <reference path="debug.ts" />

class Dealer {
	private debug: boolean;
	private game: Core;
	private index: number;
	private pos: number;
	private cards: string[];
	private values: number[];

	/**
	 * Dealer implements the dealer for Blackjack.
	 * @constructor
	 * @param {boolean} debug - Enable debug messages?
	 * @param {Core} game - Enchant.js game object.
	*/
	constructor(debug: boolean, game: Core) {
		this.debug = debug;
		this.game = game;
		this.index = 0;
		this.pos = 225;
		this.cards = new Array<string>();
		this.values = new Array<number>();
	}

	/**
	 * Calculate the total value of dealer's held cards.
	 * @returns {number} Total value of dealer's cards.
	*/
	public calcTotal(): number {
		this.values.sort(function(a: number, b: number) { return b - a });
		var total: number = 0;
		for (var i: number = 0; i < this.values.length; i++) {
			var v: number = this.values[i];
			if(v == 1) {
				if ((total + 11) <= 21) v = 11;
				else if ((total + 11) > 21) v = 1;
			}
			total += v;
		}
		return total;
	}

	/**
	 * Dealer hits.
	 * @param {Cards} cards - Game cards.
	 * @returns {Card} Dealer's drawn card.
	*/
	private hit(cards: Cards): Card {
		this.index++;
		this.pos += 90;
		var card: string = cards.draw();
		this.cards.push(card);
		this.values.push(cards.getValue());
		Debug.emit(this.debug, 'Dealer hits.');
		Debug.emit(this.debug, "Dealer gets ${card}");
		return new Card(Card.getImage(card), this.pos, 10, this.game);
	}

	/**
	 * Dealer stands.
	*/
	private stand(): void {
		Debug.emit(this.debug, 'Dealer stands.');
	}

	/**
	 * Dealer shuffles.
	 * @param {Cards} cards - Game cards to shuffle.
	*/
	public shuffle(cards: Cards): void {
		if (cards.getPlayed() == 0 || cards.getPlayed() >= 45) {
			SoundEffects.play(this.game, 'shuffle');
			Debug.emit(this.debug, '-------------------------------------------------------');
			Debug.emit(this.debug, 'Dealer is shuffling cards...');
			Debug.emit(this.debug, '-------------------------------------------------------');
			cards.shuffle();
		} 
	}

	/**
	 * Dealer deals.
	 * @param {Cards} cards - Game cards.
	 * @returns {string[]} Player's cards.
	*/
	public deal(cards: Cards): string[] {
		SoundEffects.play(this.game, 'deal');
		var dealt: string[] = new Array<string>();
		var i: number = 1;
		Debug.emit(this.debug, '-------------------------------------------------------');
		Debug.emit(this.debug, 'Dealer is dealing cards for a new game...');
		Debug.emit(this.debug, '-------------------------------------------------------');
		while(i <= (2 * 2)) {
			dealt.push("${cards.draw()}:${cards.getValue()}");
			i++;
		}
		i = 0;
		while(i < 2) {
			var cv: string[] = dealt[i].split(':');
			this.cards.push(cv[0]);
			this.values.push(parseInt(cv[1]));
			i++;
		}
		Debug.emit(this.debug, '\nDealer has:');
		Debug.emit(this.debug, '[**]${this.cards[1]}');
		return [dealt[2], dealt[3]];
	}

	/**
	 * Determine if dealer has Blackjack.
	 * @returns {boolean} Does dealer have Blackjack?
	*/
	public hasBlackjack(): boolean {
		var blackjack: boolean = false;
		if(this.calcTotal() == 21) {
			Debug.emit(this.debug, 'Dealer has Blackjack!');
			blackjack = true;
		}
		return blackjack;
	}

	/**
	 * Determien if dealer is bust.
	 * @returns {boolean} Is dealer bust?
	*/
	public isBust(): boolean {
		var bust: boolean = false;
		if(this.calcTotal() > 21) {
			Debug.emit(this.debug, 'Dealer is bust!');
			bust = true;
		}
		return bust;
	}

	/**
	 * Dealer responds to player action (e.g. a hit or stand).
	 * @param {Cards} cards - Game cards.
	 * @returns {Card[]} Cards returned.
	*/
	public respond(cards: Cards): Card[] {
		this.showCards();
		var responding: boolean = true;
		var response_cards: Card[] = new Array<Card>();
		while(responding) {
			var total: number = 0;
			while(total <= 18) {
				total = this.calcTotal();
				if (total == 16) {
					if (Math.floor(Math.random() * 6) >= 3)
						response_cards.push(this.hit(cards)); // Take risk.
					else
						this.stand(); // Play it safe.
				}
				else if (total >= 17) {
					this.stand();
					responding = false;
					break;
				}
				else
					response_cards.push(this.hit(cards));
			}
		}
		return response_cards;
	}

	/**
	 * Show dealer's cards.
	 * @returns {number} Total value of dealer's cards.
	*/
	public showCards(): number {
		this.index = 0;
		this.pos = 225;
		var cards: string = '';
		for (var i = 0; i < this.cards.length; i++) {
			cards += this.cards[i];
		}
		Debug.emit(this.debug, '\nDealer has:');
		Debug.emit(this.debug, "${cards} --> ${this.calcTotal()}");
		return this.calcTotal();
	}

	/**
	 * Dealer receives cards.
	 * @returns {Card[]} Dealer's received cards.
	*/
	public receiveCards(): Card[] {
		this.index++;
		var cardA = new Card(Card.getImage('c'), this.pos, 10, this.game);
		this.index++;
		this.pos += 90;
		var cardB = new Card(Card.getImage(this.cards[1]), this.pos, 10, this.game);
		return [cardA, cardB];
	}

	/**
	 * Dealer reveals first card.
	 * @returns {Card} Revealed first card.
	*/
	public revealFirstCard(): Card {
		SoundEffects.play(this.game, 'reveal');
		return new Card(Card.getImage(this.cards[0]), 225, 10, this.game);
	}
}
